Internal Affairs Part 1: Fax Machines!

ike709 opened the pull request:
:cl: ike709
add: Added fax machines. Now people can send faxes to other fax machines on the ship, or to Central Command. Only the QM, Captain, Bridge Officers, Lawyer, MaA, and Heads of Staff can use it. Internal Affairs Agents can use it when I finish adding them. Fax machines aren't mapped in yet.
/:cl:

Ported from baycode.

I also unfucked photocopier.dm. It had fucked up whitespace from the rebase.

I'll map them in with a separate PR after this is merged to reduce the chances of conflicts.

I'm sure all of this is shitcode and ninja will never merge it.

A note in case someone else ever maps these in for some reason: You need to edit the department variable on each fax machine you add. The department var is the name the fax machine will have on the list of fax machines you can send a fax to.

ike709 commented:
Well shit. I still have map edits for when I mapped fax machines in and later deleted them. Oh well, it's fine as long as nobody creates map conflicts.

ike709 commented on ftl13.dme:
```diff
@@ -733,10 +733,10 @@
-#include "code\game\objects\effects\decals\cleanable\humans.dm"
-#include "code\game\objects\effects\decals\cleanable\misc.dm"
-#include "code\game\objects\effects\decals\cleanable\robots.dm"
+#include "code\game\objects\effects\decals\Cleanable\aliens.dm"
```
what the fuck

ninjanomnom replied:
make sure your actual directory is lower case

ninjanomnom commented on code/modules/paperwork/faxmachine.dm:
```diff
@@ -0,0 +1,230 @@
+	use_power(200)
+
+	var/success = 0
+	for(var/obj/machinery/photocopier/faxmachine/F in GLOB.allfaxes)
```
if allfaxes only contains fax maxhines this needs to be a typeless loop

ike709 replied:
I think I did this right?

ike709 commented:
@ninjanomnom blame baycode

ike709 commented on code/modules/paperwork/faxmachine.dm:
```diff
@@ -0,0 +1,230 @@
+
+	user << browse(dat, "window=copier")
+	onclose(user, "copier")
+	return
```
Unhelpful.

ninjanomnom replied:
im telling you to completely redo this proc to be more in line with other uis

ike709 replied:
Later.

ike709 commented on code/modules/admin/admin.dm:
```diff
@@ -842,3 +842,83 @@
+
+	spawn(100)
+		qdel(P)
+		faxreply = null
```
Unhelpful.

ninjanomnom replied:
you should know exactly what's wrong here

ike709 replied:
I assume you're referring to the spawn()?

ninjanomnom replied:
If P is a temporary object then it should be deleted in the same namespace that it's created in. Deleting P in the proc it's passed to is unintuitive.

Spawn is also a problem here but a relatively minor one.

ike709 replied:
Later.

Qustinnus commented on cfg/admin.txt:
```diff
@@ -1,4 +1,5 @@
 optimumtact role=admin
+newsta role=admin
```
it happend to me too hahahahahahahhahahahahhahaha

AuroranAI replied:
wait
did we just fucking copy TG's admin list during the rebase

ike709 replied:
we don't use the config file admin list anyways

ike709 commented on code/modules/admin/admin.dm:
```diff
@@ -842,3 +842,83 @@
+
+/datum/admins/var/obj/item/weapon/paper/admin/faxreply // var to hold fax replies in
+
+/datum/admins/proc/faxCallback(var/obj/item/weapon/paper/admin/P, var/obj/machinery/photocopier/faxmachine/destination)
```
I don't know how to do this.

ike709 commented on code/modules/admin/admin.dm:
```diff
@@ -842,3 +842,83 @@
+	var/obj/item/rcvdcopy
+	rcvdcopy = destination.copy(P)
+	rcvdcopy.loc = null //hopefully this shouldn't cause trouble
+	GLOB.adminfaxes += rcvdcopy
```
I don't know how to do this.

ike709 commented on code/modules/paperwork/adminpaper.dm:
```diff
@@ -0,0 +1,120 @@
+	text += "Failure to comply is a breach of regulation and may be prosecuted to the fullest extent of the law, where applicable."
+	text += "</font>"
+
+	footer = text
```
I don't know how to do this.

ike709 commented on code/modules/paperwork/faxmachine.dm:
```diff
@@ -0,0 +1,230 @@
+
+/obj/machinery/photocopier/faxmachine/Topic(href, href_list)
+	if(href_list["send"])
+		if(copy)
```
I think I did this right?

ike709 commented on code/modules/paperwork/faxmachine.dm:
```diff
@@ -0,0 +1,230 @@
+
+	if(!admin_departments)
+		admin_departments = list("Central Command", "Centcomm Bureau of Bureaucracy", "Centcomm Supply", "Central Command Internal Affairs")
+	GLOB.allfaxes += src
```
I think I did this right?

ike709 commented on code/modules/paperwork/faxmachine.dm:
```diff
@@ -0,0 +1,230 @@
+	var/obj/item/weapon/card/id/scan = null // identification
+	var/authenticated = 0
+	var/sendcooldown = 0 // to avoid spamming fax messages
+	var/department = "Unknown" // our department
```
I think I did this right?

ike709 commented on code/modules/paperwork/adminpaper.dm:
```diff
@@ -0,0 +1,120 @@
+	interactions += "<A href='?src=\ref[src];toggleheader=1'>Toggle Header</A> "
+	interactions += "<A href='?src=\ref[src];togglefooter=1'>Toggle Footer</A> "
+	interactions += "<A href='?src=\ref[src];clear=1'>Clear page</A> "
+	interactions += "</center>"
```
I think I did this right?

ike709 commented:
@ninjanomnom I've addressed all of your concerns so far with either "I think I did this right, I don't know how to do this, or I'll do it later because I ran out of time"

Please look at the stuff I have done, explain how to do the stuff I don't know, and complain about anything else that is shitcode.

ninjanomnom commented on code/modules/admin/admin.dm:
```diff
@@ -842,3 +842,83 @@
+	set desc = "Sends a fax to this machine"
+	var/department = input("Choose a fax", "Fax") as null|anything in GLOB.all_departments
+	for(var/obj/machinery/photocopier/faxmachine/sendto in GLOB.all_faxes)
+		if(sendto.department == department)
```
if departments can only have 1 fax machine then make this an associative list. You can do `P.destination = GLOB.all_faxes[department]` then.

Regardless, this should be an untyped loop since I assume GLOB.all_faxes only contains fax machines.

ninjanomnom commented on code/modules/admin/admin.dm:
```diff
@@ -842,3 +842,83 @@
+				src = usr.client.holder
+			if (!istype(src,/datum/admins))
+				to_chat(usr, "Error: you are not an admin!")
+				return
```
Why are there 2 if checks for the same thing?

ike709 replied:
I think I did this right?

ninjanomnom commented on code/modules/admin/admin.dm:
```diff
@@ -842,3 +842,83 @@
+		var/image/stampoverlay = image('icons/obj/bureaucracy.dmi')
+		var/{x; y;}
+		x = rand(-2, 0)
+		y = rand(-1, 2)
```
Please don't initialize variables like this.

ike709 replied:
I think I did this right?

ninjanomnom commented on code/modules/admin/topic.dm:
```diff
@@ -1743,6 +1743,33 @@
 		to_chat(H, "You hear something crackle in your ears for a moment before a voice speaks.  \"Please stand by for a message from your benefactor.  Message as follows, agent. [input].  Message ends.\"")
 
+	else if(href_list["AdminFaxView"])
+		var/obj/item/fax = locate(href_list["AdminFaxView"])
```
You need a better way of getting the fax than a locate like this. Fetch it from a known list instead.

ninjanomnom commented on code/modules/admin/topic.dm:
```diff
@@ -1743,6 +1743,33 @@
+
+	else if(href_list["FaxReply"])
+		var/mob/sender = locate(href_list["FaxReply"])
+		var/obj/machinery/photocopier/faxmachine/fax = locate(href_list["originfax"])
```
More locates that need to not exist.

ninjanomnom commented on code/modules/paperwork/adminpaper.dm:
```diff
@@ -0,0 +1,121 @@
+//Adminpaper - it's like paper, but more adminny! Used for sending faxes.
+/obj/item/weapon/paper/admin
```
Make a datum that formats a piece of paper correctly instead of having a special piece of paper

ninjanomnom commented on code/modules/paperwork/adminpaper.dm:
```diff
@@ -0,0 +1,121 @@
+
+
+/obj/item/weapon/paper/admin/proc/generate_interactions()
+	var/temp_interactions = null
```
no need for the ` = null` part

ike709 replied:
I think I did this right?

Qustinnus commented:
@ninjanomnom you broke ike
